Testnet/NFTStore/NFTStore/NFTStore.cs

```diff
@@ -0,0 +1,155 @@
+﻿using Stratis.SmartContracts;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+[Deploy]
+public class NFTStore : SmartContract
+{
+    public SaleInfo GetSaleInfo(Address contract, ulong tokenId) => State.GetStruct<SaleInfo>($"SaleInfo:{contract}:{tokenId}");
+    private void SetSaleInfo(Address contract, ulong tokenId, SaleInfo value) => State.SetStruct($"SaleInfo:{contract}:{tokenId}", value);
+    private void ClearSaleInfo(Address contract, ulong tokenId) => State.Clear($"SaleInfo:{contract}:{tokenId}");
+
+    public NFTStore(ISmartContractState state)
+        : base(state)
+    {
+        EnsureNotPayable();
+    }
+
+    /// <summary>
+    /// Ensure the token is approved on the non fungible contract for store contract as pre-condition.
+    /// </summary>
+    public void Sale(Address contract, ulong tokenId, ulong price)
+    {
+        EnsureNotPayable();
+
+        Assert(price > 0, "Price should be higher than zero.");
+
+        var tokenOwner = GetOwner(contract, tokenId);
+
+        Assert(tokenOwner != Address, "The token is already on sale.");
+
+        EnsureCallerCanOperate(contract, tokenOwner);
+
+        TransferToken(contract, tokenId, tokenOwner, Address);
+
+        SetSaleInfo(contract, tokenId, new SaleInfo { Price = price, Seller = tokenOwner });
+
+        Log(new TokenOnSaleLog { Contract = contract, TokenId = tokenId, Price = price, Seller = tokenOwner });
+    }
+
+    public void Buy(Address contract, ulong tokenId)
+    {
+        var saleInfo = GetSaleInfo(contract, tokenId);
+
+        EnsureTokenIsOnSale(saleInfo);
+
+        Assert(Message.Value == saleInfo.Price, "Transferred amount is not matching exact price of the token.");
```

Comment:
Would it be better to allow for accidental overpayment and return change?  Saves users from a failed transaction from a perfectly valid mistake. 

```
Assert(Message.Value >= saleInfo.Price);
var change = Message.Value - saleInfo.Price;
Transfer(change, to);
```

Reply:
Users will not going to call these methods directly. We will built a transaction from user interface. Also there are cases that remaining coins may not to worth transfer back because it is very small or even returning the amount can fail so we have to introduce withdraw pattern.